feat(languagedetection): use `dd_language_detected` if available

stanistan opened the pull request:
### What does this PR do?

This PR adds a privileged detector that uses a file set in `memfd` if it is available. The injection mechanism [will write a file to memfd](https://github.com/DataDog/auto_inject/blob/13fb55691332a3adeb73ebd9859bc559c493cc57/src/libinject.c#L442-L454) if a language is detected, and the file will be in `dd_language_detected` [source](https://github.com/DataDog/auto_inject/blob/13fb55691332a3adeb73ebd9859bc559c493cc57/src/include/dd/memfd.h).

### Motivation

As we mature kubernetes auto-instrumentation, we want to leverage language detection to minimize the number of init containers we load at pod startup time. While having this mechanism might not help the _first_ time the process is running, but will help on subsequent pods.

### Describe how you validated your changes

- `apm-inject`
  - this version (`ghcr.io/datadog/apm-inject:6babf1ba57cd2b1ca5943c99b5eab9ed653529a6`) supports writes to memfd
  - this version (`ghcr.io/datadog/apm-inject:0.29.0`) does not

I'm using https://github.com/DataDog/k8s-ssi-v2-testing to set up two deployments for an application and the built image from this PR.

- agent: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/793848449
- cluster-agent: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/793848468
- Both re-published [here](https://hub.docker.com/layers/ddstanistan/cluster-agent/v54882008-ef7499ca-amd64) and [here](https://hub.docker.com/layers/ddstanistan/agent/v54882008-ef7499ca-7-amd64).

##### Manual 

On an agent container we can run this script which looks for the memfd file and makes sure that it matches what's in the workload-list.

```sh
# get a process with the memfd associated with it
filename=$((ls -lRa /proc/*/fd/* 2>/dev/null) | grep 'memfd:dd_language_detected' | awk '{ print $9 }')
pid=$(echo $filename | awk -F/ '{ print $3 }')

# output the written file as well as what's in workload
agent workload-list | grep $pid -A4
echo "/memfd:dd_language_detected: $(cat $filename)"
```

##### Automated

- added unit tests for memfd behavior when its enabled.

###### Testing Log

<img width="913" alt="Screenshot 2025-02-04 at 3 04 32 PM" src="https://github.com/user-attachments/assets/461df335-7a68-42f3-8794-f7c41ad1fe93" />

```sh
(ls -lRa /proc/*/fd/* 2>/dev/null) | grep 'memfd:dd_language_detected' | awk '{ print $9 }' | xargs cat
```

<img width="998" alt="Screenshot 2025-02-04 at 3 12 41 PM" src="https://github.com/user-attachments/assets/5bbe3a20-ffc6-437a-a912-d12f2488461e" />

This means the file is being written, and without any lang detection we're not doing anything at the deployment.

We can also find the process in workloads being correctly tagged.

<img width="677" alt="Screenshot 2025-02-04 at 3 14 53 PM" src="https://github.com/user-attachments/assets/6a72848e-6314-4c58-9c7a-2a63713b787f" />

Then the deployment gets annotated.

<img width="456" alt="Screenshot 2025-02-04 at 3 18 30 PM" src="https://github.com/user-attachments/assets/5b4e1708-f859-4167-abd8-1957505a17b6" />

delete the pod, the next one has only the js container.

<img width="1000" alt="Screenshot 2025-02-04 at 3 24 10 PM" src="https://github.com/user-attachments/assets/1b6fb0e6-9529-4e2c-8b86-277a0039ece4" />

By default the js app uses `node` which will match langdetection outside of memfd, but we can run something else `cp $(which node) app-runner && ./app-runner index.js`. And this should give us the same behavior.

<img width="423" alt="Screenshot 2025-02-04 at 3 37 48 PM" src="https://github.com/user-attachments/assets/8365aada-780b-455f-bad9-0063f4e400e6" />

<img width="651" alt="Screenshot 2025-02-04 at 3 42 19 PM" src="https://github.com/user-attachments/assets/0cdb43b6-f1d4-4d9e-ae8f-38ef67064122" />

- After turning on `system-probe` and `language-detection` by setting up env vars manually on the helm chart: ✅ 

<img width="648" alt="Screenshot 2025-02-04 at 5 26 09 PM" src="https://github.com/user-attachments/assets/e7ea90c7-860d-4651-981d-be638f1db170" />

### Additional Notes

- There is test coverage for different languages having their `language_detection` memfd files present in the auto-inject repository.

- This factors out memfd code to `util/kernel` since it's also used in servicediscovery (even though auto-inject has not wired this up yet).

stanistan committed:
feat(languagedetection): use dd-injector-detected-langauge if available.

stanistan committed:
add package docs

vitkyrka commented:
> system-probe is not enabled by default, you have to be doing discovery, or network monitoring, etc for this to be on, so we're not getting as much from this as I had hoped.
> not only that, but there is no option to turn on system-probe language-detection that is enabled for users, so this change is basically dark.

When service discovery is enabled (it's on when USM is enabled currently), it will do language detection in system-probe so this code will be exercised in that case.

stanistan committed:
remove unrelated changes

stanistan committed:
move reading memfd file to util/kernel

stanistan commented on pkg/languagedetection/internal/detectors/injector.go:
```diff
@@ -0,0 +1,131 @@
+	fdsPath := kernel.HostProc(strconv.Itoa(pid), "fd")
+	// quick path, the shadow file is the first opened file by the process
+	// unless there are inherited fds
+	path := filepath.Join(fdsPath, "3")
```
I think the language file will be first, and then we will add the other one, the implementation will work either way, and I was thinking that we could add a "fd hint" in the future for optimizations if needed.

stanistan commented on pkg/languagedetection/internal/detectors/injector.go:
```diff
@@ -0,0 +1,131 @@
+	pid := proc.GetPid()
+	path, found := findInjectorFile(int(pid))
+	if !found {
+		return model.Language{}, errors.New("no injector file found")
```
👍 Addressed in the move to the common package.

stanistan commented on pkg/languagedetection/detector.go:
```diff
@@ -126,7 +126,7 @@ func DetectLanguage(procs []languagemodels.Process, sysprobeConfig model.Reader)
 			}
 
-			if lang.Name != languagemodels.Unknown {
+			if !lang.IsUnknown() {
```
Removed these.

stanistan committed:
fix lint

stanistan committed:
w/ tests for injector language detection

stanistan commented on pkg/languagedetection/internal/detectors/injector.go:
```diff
@@ -0,0 +1,131 @@
+//     source: https://github.com/DataDog/auto_inject/blob/13fb55691332a3adeb73ebd9859bc559c493cc57/src/include/dd/memfd.h#L3
+const memfdLanguageDetectedFileName = "/memfd:dd_language_detected"
+
+// NewInjectorDetector is a language Detector that uses
```
Added tests!

stanistan committed:
fix lint

stanistan committed:
with release notes

drichards-87 commented on releasenotes/notes/injector-language-detection-e2691e28c6273286.yaml:
```diff
@@ -0,0 +1,6 @@
+enhancements:
+  - |
+    Added a new language detector that uses the apm-inject propagated
+    language if it is present.
```
```suggestion
    language, if available.
```

drichards-87 reviewed (approved):
Left a very small suggestion from Docs and approved the PR.

stanistan committed:
Update releasenotes/notes/injector-language-detection-e2691e28c6273286.yaml

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

vitkyrka commented on pkg/util/kernel/proc.go:
```diff
@@ -107,3 +108,91 @@ func GetProcessEnvVariable(pid int, procRoot string, envVar string) (string, err
+	}
+	fdDir, err := os.Open(fdsPath)
+	if err != nil {
+		// log.Warnf("failed to open %s: %s", fdsPath, err)
```
Were these commented-out lines kept on purpose?

brycekahle commented on pkg/util/kernel/proc.go:
```diff
@@ -107,3 +108,91 @@ func GetProcessEnvVariable(pid int, procRoot string, envVar string) (string, err
+		return nil, err
+	}
+
+	if len(data) == memFdMaxSize {
```
Because of this, I don't think you can ever read exactly `memFdMaxSize` without error. There is no way to differentiate between the `LimitReader` limiting you and reading the exact max.

betterengineering replied:
Hey @brycekahle! I'm going to attempt to shephard this PR along while Stan is out on PTO. I pushed a solution where we read max+1 bytes to compare against, let me know what you think.

betterengineering committed:
fix bug in mamFdMaxSize

This commit resolves an issue where reading a file of exactly the max
size would be an error condition. It does so by reading an extra byte
and sending an error if the size is larger then the max size.

betterengineering commented on pkg/util/kernel/proc.go:
```diff
@@ -107,3 +108,91 @@ func GetProcessEnvVariable(pid int, procRoot string, envVar string) (string, err
+	}
+	fdDir, err := os.Open(fdsPath)
+	if err != nil {
+		// log.Warnf("failed to open %s: %s", fdsPath, err)
```
```suggestion
```

betterengineering committed:
Remove commented out log line

betterengineering commented:
/merge
